File: csharp/1695.maximum-erasure-value.cs
/*
 * @lc app=leetcode id=1695 lang=csharp
 *
 * [1695] Maximum Erasure Value
 */

using System;
using System.Collections.Generic;

// @lc code=start
public class Solution {
    public int MaximumUniqueSubarray(int[] nums) {
        int acc = 0, p = 0, ret = 0, n = nums.Length;
        var seen = new HashSet<int>();
        for (int i = 0; i < n; ++i) {
            int curr = nums[i];
            if (!seen.Add(curr)) {
                while (nums[p] != curr) {
                    seen.Remove(nums[p]);
                    acc -= nums[p++];
                }
                ++p;
            } else {
                acc += curr;
            }
            ret = Math.Max(acc, ret);
        }
        return ret;
    }
}
// @lc code=end

File: java/1695.maximum-erasure-value.java
/*
 * @lc app=leetcode id=1695 lang=java
 *
 * [1695] Maximum Erasure Value
 */

// @lc code=start
class Solution {
    public int maximumUniqueSubarray(int[] nums) {
        boolean[] contains = new boolean[100001];
        int n = nums.length, l = 0, acc = 0, ret = 0;
        for (int r = 0; r < n; ++r) {
            int x = nums[r];
            if (contains[x]) {
                while (nums[l] != x) {
                    contains[nums[l]] = false;
                    acc -= nums[l++];
                }
                ++l;
            } else {
                acc += x;
                contains[x] = true;
            }
            ret = Math.max(ret, acc);
        }
        return ret;
    }
}
// @lc code=end

